### PR TITLE
ES6: convert to ES6 classes from extendable-base

### PR DESCRIPTION
**Pull request by demmer**
Run the script at https://gist.github.com/demmer/562cfc2343cc9d049b69
to convert all the extendable-base classes to ES6 classes.

This also required a rework of how the error classes were implemented.

### CODE REVIEW

**Comment by demmer** [comment]
@mstemm

File: lib/errors.js [hunk 1]
```diff
@@ -28,140 +28,135 @@ var BaseError = Base.inherits(Error, {
...
-});
+class UnknownError extends BaseError {
+    constructor(info) {
+        super('UnknownError', 'JS-UNKNOWN-ERROR', 500, info);
+    }
+}
 
 
-// All errors from the juttle compiler and runtime are grouped as this
+// All var from the juttle compiler and runtime are grouped as this
```
Comment (mstemm) [inline lib/errors.js]: Is this just the script gone wild?

Comment (demmer) [inline lib/errors.js]: Ha. I think that one was my manual search-and-replace failure. Will update.

File: lib/errors.js [hunk 2]
```diff
@@ -28,140 +28,135 @@ var BaseError = Base.inherits(Error, {
...
+    return (err instanceof UnknownError ||
+            err instanceof JuttleError ||
+            err instanceof BundleError ||
+            err instanceof JobNotFoundError ||
+            err instanceof FileNotFoundError ||
+            err instanceof FileAccessError ||
+            err instanceof DirectoryNotFoundError ||
+            err instanceof DirectoryAccessError ||
+            err instanceof InvalidPathError ||
+            err instanceof TimeoutError);
```
Comment (demmer) [inline lib/errors.js]: Not strictly related but couldn't this be simply `err instanceof BaseError`?

File: lib/errors.js [hunk 3]
```diff
@@ -28,140 +28,135 @@ var BaseError = Base.inherits(Error, {
...
-    code: 'JS-INVALID-PATH-ERROR',
-    status: 400
-});
-
-errors.TimeoutError = BaseError.extend({
-    name: 'TimeoutError',
-    code: 'JS-TIMEOUT-ERROR',
-    status: 408
-});
+class JuttleError extends BaseError {
```
Comment (mstemm) [inline lib/errors.js]: This means the error classes won't be exported. I guess that's further incentive to use the creator functions instead, but I think we should be consistent between juttle-service and juttle once juttle is converted to use classes.

Comment (demmer) [inline lib/errors.js]: Yeah I know... As it stands the conventions are a little different.

I'd be ok with adding the classes to exports if we need them.

File: lib/errors.js (hunk 2 above)
Comment (mstemm) [inline lib/errors.js]: I think it could. I wasn't certain instanceof worked properly with extendable-base, but it certainly does for classes.

**Comment by mstemm** [comment]
I'd say also collapse the instanceof checks to just BaseError as you suggested, and then lgtm!

**Comment by demmer** [comment]
@mstemm I needed to rework how the http status was handled so it wouldn't end up in the returned object. Take a look at the last commit and let me know what you think.

**Comment by mstemm** [comment]
I didn't understand why you had to keep the status out of the object instances, but I'm equally happy with the updated implementation, so lgtm.

**Comment by demmer** [comment]
I didn't _have_ to keep it out but it changed the format on the wire and it
seemed redundant to include the http status in the JSON body of the errors.


On Thu, Feb 11, 2016 at 3:52 PM, Mark Stemm <notifications@github.com>
wrote:

> I didn't understand why you had to keep the status out of the object
> instances, but I'm equally happy with the updated implementation, so lgtm.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/juttle/juttle-service/pull/41#issuecomment-183115534>.
>



-- 
Michael Demmer
Jut - CTO / VP Engineering
demmer@jut.io

**Pull request merged**